planner/core/planbuilder.go

```diff
@@ -2414,6 +2414,11 @@ func (b *PlanBuilder) buildSimple(ctx context.Context, node ast.StmtNode) (Plan,
 			b.visitInfo = appendVisitInfoIsRestrictedUser(b.visitInfo, b.ctx, user, "RESTRICTED_USER_ADMIN")
 		}
 	case *ast.RevokeStmt:
+		if b.ctx.GetSessionVars().CurrentDB == "" && raw.Level.DBName == "" {
+			if raw.Level.Level != ast.GrantLevelGlobal {
+				return nil, ErrNoDB
+			}
+		}
 		b.visitInfo = collectVisitInfoFromRevokeStmt(b.ctx, b.visitInfo, raw)
```

Comment:
I think it may be intended or so, check this https://github.com/pingcap/tidb/blob/92ddceb6a4ab1c932fd0390a3a4d710a76da1f42/planner/core/planbuilder.go#L2489-L2498.

If you ensure that there is no problem, I would suggest you put the check into `collectVisitInfoFromRevokeStmt` and `collectVisitInfoFromGrantStmt`.

Reply:
Yes, I agree it might be better to put the checks in one place.

Btw; I originally fixed this in https://github.com/pingcap/tidb/pull/23946/files but it looks like you have a more correct fix because the check I fixed does not account for `CurrentDB` being empty.

Reply:
@morgo yes，i find your fix.  https://github.com/pingcap/tidb/issues/25755#issuecomment-868165376
this  pr is a make up. the case is CurrentDB being empty , grant/revoke on * to/from user, behave not compatible with mysql.